internal/database/sqlcommon/message_sql.go

```diff
@@ -56,15 +56,86 @@ var (
 	}
 )
 
-func (s *SQLCommon) UpsertMessage(ctx context.Context, message *fftypes.Message, allowExisting, allowHashUpdate bool) (err error) {
+func (s *SQLCommon) attemptMessageUpdate(ctx context.Context, tx *txWrapper, message *fftypes.Message) (int64, error) {
+	return s.updateTx(ctx, tx,
+		sq.Update("messages").
+			Set("cid", message.Header.CID).
+			Set("mtype", string(message.Header.Type)).
+			Set("author", message.Header.Author).
+			Set("key", message.Header.Key).
+			Set("created", message.Header.Created).
+			Set("namespace", message.Header.Namespace).
+			Set("topics", message.Header.Topics).
+			Set("tag", message.Header.Tag).
+			Set("group_hash", message.Header.Group).
+			Set("datahash", message.Header.DataHash).
+			Set("hash", message.Hash).
+			Set("pins", message.Pins).
+			Set("state", message.State).
+			Set("confirmed", message.Confirmed).
+			Set("tx_type", message.Header.TxType).
+			Set("batch_id", message.BatchID).
+			Where(sq.Eq{
+				"id":   message.Header.ID,
+				"hash": message.Hash,
+			}),
+		func() {
+			s.callbacks.OrderedUUIDCollectionNSEvent(database.CollectionMessages, fftypes.ChangeEventTypeUpdated, message.Header.Namespace, message.Header.ID, -1 /* not applicable on update */)
+		})
+}
+
+func (s *SQLCommon) attemptMessageInsert(ctx context.Context, tx *txWrapper, message *fftypes.Message) (err error) {
+	message.Sequence, err = s.insertTx(ctx, tx,
+		sq.Insert("messages").
+			Columns(msgColumns...).
+			Values(
+				message.Header.ID,
+				message.Header.CID,
+				string(message.Header.Type),
+				message.Header.Author,
+				message.Header.Key,
+				message.Header.Created,
+				message.Header.Namespace,
+				message.Header.Topics,
+				message.Header.Tag,
+				message.Header.Group,
+				message.Header.DataHash,
+				message.Hash,
+				message.Pins,
+				message.State,
+				message.Confirmed,
+				message.Header.TxType,
+				message.BatchID,
+			),
+		func() {
+			s.callbacks.OrderedUUIDCollectionNSEvent(database.CollectionMessages, fftypes.ChangeEventTypeCreated, message.Header.Namespace, message.Header.ID, message.Sequence)
+		})
+	return err
+}
+
+func (s *SQLCommon) UpsertMessage(ctx context.Context, message *fftypes.Message, optimization database.UpsertOptimization) (err error) {
 	ctx, tx, autoCommit, err := s.beginOrUseTx(ctx)
 	if err != nil {
 		return err
 	}
 	defer s.rollbackTx(ctx, tx, autoCommit)
 
-	existing := false
-	if allowExisting {
+	// This is a performance critical function, as we stream data into the database for every message, in every batch.
+	//
+	// First attempt the operation based on the optimization passed in.
+	// The expectation is that this will practically hit of the time, as only recovery paths
```

Comment:
nit: this sentence feels like it's missing a word (same with the other places it was copied)